src/core/dev_api/tensor_data_accessor.hpp

```diff
@@ -0,0 +1,83 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include "openvino/core/node.hpp"
+#include "tensor_data_adapter.hpp"
+
+namespace ov {
+
+/** @brief Function object type to get data from object container as tensor data adapter. */
+using tensor_data_accessor_func_t = std::function<ITensorDataAdapter::UPtr(size_t)>;
+
+/**
+ * @brief Tensor data accessor functor.
+ *
+ * Creates the tensor data adapter if found in tensor container.
+ * This accessor not take ownership of tensors container.
+ * Supports following containers:
+ * - ov::TensorVector
+ * - ov::HostTensorVector
+ * - std::map<size_t, ov::HostTensorPtr>
+ *
+ * @tparam TContainer Type of tensor container.
+ */
+template <class TContainer>
+struct OPENVINO_API TensorAccessor {
```

Comment:
Just wondering, why we cannot simply introduce such accessor interface and every plugin just implement it on its side? So we just pass the accessor to the infer call.

Reply:
I assumed this interface is specified by `tensor_data_accessor_func_t` which define function object which can be passed.
So in plugin side it can be implemented by anything (not bind to specific class) like lambda, functor, function which satisfy this function object.
The class `ov::TensorAccessr` is to provide this function object mainly for tensors (ov::Tensor, HostTensor) where I specified specialization for `operator()` for different kind of tensors and containers which hold them. It looks like these specializations cover all cases from core perspective. They are in dev API because they are shared between libraries.

As I don't know what exactly is required by plugin I left it open for implementation. Plugin can have own class or maybe will be enough to add in plugin own `tensor adapter class` and then specialization for `operator()` and re-use `make_tensor_accessor` helper to create it as in tile operator [here](https://github.com/openvinotoolkit/openvino/blob/b98e9504e4a919977871eceb962ad4037b32f05a/src/core/src/op/tile.cpp#L59) and [here](https://github.com/openvinotoolkit/openvino/blob/b98e9504e4a919977871eceb962ad4037b32f05a/src/core/src/op/tile.cpp#L84)

Reply:
I got the idea, thank you. But, this functional approach being more flexible may have some performance drawbacks. The thing is, the shape infer operations should be as quick as possible. Also some of them are really simple, like merging two vectors of 4 elements each. Though, any ongoing overheads might be really big in comparison with the shape infer algorithm itself. For example, passing the functor as a value may involve additional copy of internal structures, even though this is not a big problem from the interface stand point, since it is always possible to say that the user should implement more lightweight functors, it is still better to avoid such copy. Another point is the return type, which is a unique_ptr. That means that most likely we should allocate dynamic memory to store the adapter every access request call. Which requires `malloc`, which is not cheap.
But, in opposite, we can take a reference to some accessor abstract interface, which always return a tensor of specific type, so all the internal optimizations are on the plugin side and we do not impose any additional overheads on the interface level. That means on the plugin side I can instantiate such adapter implementation object only once, as well as all the auxiliary structures and pass it as a reference every infer call. Also the accessor implementation may return a reference to some thin wrapper over the internal tensor implementation that provides direct access to the internal data with minimal overhead (at lest in the case of CPU plugin since the data are stored in the same DRAM).

Reply:
Regarding passing accessor by reference and avoid using unique_ptr I can re-factor it to not use.

The current tensor accessor returns `ITensorDataAdapter` which can store tensor pointer or memory pointer and the `data()` member function return pointer to data. There are additional member functions to get element type and size to jump over data correctly. Do you think about something thinner?

In case of GPU plugin, this example adapter holds also lock to data to be sure that during shape inference data are locked and released after. If accessor return reference not sure how to hold correctly this lock and provide correct lifetime.

For `ov:tensor` and `host tensor` it warp tensors to unify the interface between tensors.

Reply:
> The current tensor accessor returns ITensorDataAdapter which can store tensor pointer or memory pointer and the data() member function return pointer to data. There are additional member functions to get element type and size to jump over data correctly. Do you think about something thinner?

Honestly it completely depends on how it is used, if this interface is sufficient, then OK.

> In case of GPU plugin, this example adapter holds also lock to data to be sure that during shape inference data are locked and released after. If accessor return reference not sure how to hold correctly this lock and provide correct lifetime.

That is why I'm asking about passing the accessor instead of the factory method. In case we pass the accessor by reference, its lifetime is controlled by the plugin, so the shape infer implementation should not bother about the accessor lifespan, and, in its turn, it can return a reference to the tensor adapter, and again the lifespan is controlled by the plugin (since all the locks and auxiliary data structures are managed by the accessor). Or m.b. I'm missing something and we want to store the adapter reference beyond the `shape_infer` call and want to mange its lifetime from the shape inference subsystem?

Reply:
There is version to pass accessor via reference